middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -471,17 +731,41 @@ let close_let acc env id user_visible defining_expr
     | Some (Simple simple) ->
       let body_env = Env.add_simple_to_substitute env id simple in
       body acc body_env
-    | Some _ | None -> (
+    | None ->
       (* CR pchambart: Not tail ! *)
       let acc, body = body acc body_env in
-      match defining_expr with
-      | None -> acc, body
-      | Some defining_expr ->
-        let var = VB.create var Name_mode.normal in
-        Let_with_acc.create acc
-          (Bound_pattern.singleton var)
-          defining_expr ~body
-        |> Expr_with_acc.create_let)
+      acc, body
+    | Some defining_expr ->
+      let body_env =
+        match defining_expr with
+        | Prim (Variadic (Make_block (_, Immutable), fields), _) ->
+          let approxs =
+            List.map (Env.find_value_approximation body_env) fields
+            |> Array.of_list
+          in
+          Env.add_block_approximation body_env (Name.var var) approxs
+        | Prim (Binary (Block_load _, block, field), _) -> begin
+          match Env.find_value_approximation body_env block with
+          | Value_unknown -> body_env
+          | Closure_approximation _ -> assert false
```

Comment:
You should use `Misc.fatal_error` instead. You never know when someone is going to write code like `let f x = x + clos_var in let (_, _, clos_var) = Obj.magic f`.